api/webrtc/webrtc.go

```diff
@@ -0,0 +1,298 @@
+		return err
+	}
+
+	answer, err := c.pc.CreateAnswer(nil)
+	if err != nil {
+		return err
+	}
+
+	if err := c.pc.SetLocalDescription(answer); err != nil {
+		return err
+	}
+
+	c.pc.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			localDesc := c.pc.LocalDescription()
+			encodedDescr, err := Encode(localDesc)
+			if err != nil {
+				log.Error("WebRTC OnICECandidate error: ", err)
+				return
+			}
+			c.answer = encodedDescr
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	c.pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		dc.OnOpen(func() {
+			c.dc = dc
+
+			c.mutex.Lock()
+			c.isConnected = true
+			c.mutex.Unlock()
+
+			if NewConnection != nil {
+				go NewConnection(c, nil)
+			}
+		})
+
+		dc.OnMessage(func(msg webrtc.DataChannelMessage) {
+			var dcmsg DataChannelMessage
+			if msg.IsString {
+				dcmsg.messageType = TextMessage
+				if string(msg.Data) == PingData {
+					dc.SendText(PongData)
+					return
+				} else if string(msg.Data) == PongData {
+					if c.pongHandler != nil {
+						c.pongHandler(PongData)
+					} else {
+						log.Info("Pong handler not set")
+					}
+					return
+				}
+			} else {
+				dcmsg.messageType = BinaryMessage
+			}
+			dcmsg.data = msg.Data
+			c.OnMessage <- dcmsg
+		})
+
+		dc.OnClose(func() {
+			c.mutex.Lock()
+			c.isConnected = false
+			c.mutex.Unlock()
+		})
+
+		dc.OnError(func(err error) {
+			log.Errorf("Data Channel %v error: %s\n", dc.Label(), err.Error())
+		})
+	})
+
+	return nil
+}
+
+func (c *Peer) IsConnected() bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.isConnected
+}
+
+func (c *Peer) SetRemoteSdp(sdp string) error {
+	answer := webrtc.SessionDescription{}
+	err := Decode(sdp, &answer)
+	if err != nil {
+		return err
+	}
+	return c.pc.SetRemoteDescription(answer)
+}
+
+func (c *Peer) WriteMessage(messageType int, data []byte) (err error) {
+	if c.dc == nil {
+		return fmt.Errorf("DataChannel not available")
+	}
+	if data == nil {
+		if messageType == PingMessage {
+			return c.dc.SendText(PingData)
+		} else if messageType == PongMessage {
+			return c.dc.SendText(PongData)
+		}
+		return nil
+	}
+
+	writeResult := make(chan error)
+	go func() {
+		if messageType == TextMessage {
+			err = c.dc.SendText(string(data))
+		} else {
+			err = c.dc.Send(data)
+		}
+
+		writeResult <- err
+	}()
+
+	var dur time.Duration
+	if c.writeDeadline.After(time.Now()) {
+		dur = time.Until(c.writeDeadline)
+		select {
+		case err := <-writeResult:
+			return err
+		case <-time.After(dur):
+			return fmt.Errorf("webrtc write deadline timeout")
+		}
+	} else {
+		err := <-writeResult
+		return err
+	}
+}
+
+// WriteJSON writes the JSON encoding of v as a message.
+func (c *Peer) WriteJSON(v interface{}) error {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	return c.WriteMessage(TextMessage, b)
+}
+
+func (c *Peer) ReadMessage() (messageType int, data []byte, err error) {
+	if c.readDeadline.After(time.Now()) {
```

Comment:
The behavior is still different from tcp. To be more specific, if deadline has past (but not zero), all future read should throw error. Check the following doc from net package:
```
        // SetDeadline sets the read and write deadlines associated
	// with the connection. It is equivalent to calling both
	// SetReadDeadline and SetWriteDeadline.
	//
	// A deadline is an absolute time after which I/O operations
	// fail instead of blocking. The deadline applies to all future
	// and pending I/O, not just the immediately following call to
	// Read or Write. After a deadline has been exceeded, the
	// connection can be refreshed by setting a deadline in the future.
	//
	// If the deadline is exceeded a call to Read or Write or to other
	// I/O methods will return an error that wraps os.ErrDeadlineExceeded.
	// This can be tested using errors.Is(err, os.ErrDeadlineExceeded).
	// The error's Timeout method will return true, but note that there
	// are other possible errors for which the Timeout method will
	// return true even if the deadline has not been exceeded.
	//
	// An idle timeout can be implemented by repeatedly extending
	// the deadline after successful Read or Write calls.
	//
	// A zero value for t means I/O operations will not time out.
```

Reply:
Yes. let me modify it.